### PR TITLE
Load patching, lazy-rubygems and did_you_mean during context pre-initialization

### PR DESCRIPTION
**Pull request by eregon**

### CODE REVIEW

File: src/main/java/org/truffleruby/RubyContext.java [hunk 1]
```diff
@@ -348,6 +348,28 @@ private boolean compatibleOptions(Options oldOptions, Options newOptions) {
             return false;
         }
 
+        // Libraries loaded during pre-initialization
+
+        if (newOptions.PATCHING != oldOptions.PATCHING) {
+            Log.LOGGER.fine(notReusingContext + "loading patching is " + newOptions.PATCHING);
```
Comment (chrisseaton) [inline src/main/java/org/truffleruby/RubyContext.java]: What will this print? `loading patching is true` that's not very clear. How about `loading patching enabled` or something like that?

File: src/main/ruby/post-boot/post-boot.rb [hunk 2]
```diff
@@ -70,16 +77,30 @@
...
+    old_home = Truffle::Boot.ruby_home
+    Truffle::Boot.delay do
+      new_home = Truffle::Boot.ruby_home
+      [$LOAD_PATH, $LOADED_FEATURES].each do |array|
+        array.each do |path|
+          if path.start_with?(old_home)
+            path.replace(new_home + path[old_home.size..-1])
           end
         end
       end
```
Comment (chrisseaton) [inline src/main/ruby/post-boot/post-boot.rb]: What are you doing in this loop? Can you comment on what it achieves?

File: src/main/ruby/post-boot/post-boot.rb [hunk 3]
```diff
@@ -9,46 +9,53 @@
...
+  begin
+    require 'enumerator'
+    require 'thread'
+    require 'rational'
+    require 'complex'
+    require 'unicode_normalize'
+    if Truffle::Boot.get_option('patching')
+      Truffle::Boot.print_time_metric :'before-patching'
+      require 'truffle/patching'
+      # TODO
```
Comment (eregon) [inline src/main/ruby/post-boot/post-boot.rb]: I forgot about this TODO, fix coming.

File: src/main/java/org/truffleruby/RubyContext.java (hunk 1 above)
Comment (eregon) [reply src/main/java/org/truffleruby/RubyContext.java]: I was thinking to use a helper function to log this in a more generic way:
```
not reusing pre-initialized context: -X<option> is <new_value> (was <old_value>)
```
as it's easy to forget to update one of the 3 or 4 occurrences of the option in this manual handling.

**Comment by eregon** [comment]
> As an aside, the list of things we need to patch is growing. While it can all be figured out from the code, I think having a running list in a contrib doc would be helpful for getting a snapshot view of the process.

A list of changes would get outdated very quickly. But I can add general documentation on how context pre-initialization works and how to find out the changes (Truffle::Boot.delay/redo/preinitializing? in Ruby + RubyContext#isPreinitializing/wasPreinitialized in Java).

**Comment by eregon** [comment]
I pushed a few more commits (from 15bc28b) to fix the CI and document pre-initialization.

**Comment by eregon** [comment]
@chrisseaton @nirvdrum  Please review the new commits.

File: src/main/java/org/truffleruby/RubyContext.java (hunk 1 above)
Comment (chrisseaton) [reply src/main/java/org/truffleruby/RubyContext.java]: Logger helper functions have a problem - they ruin the meta information you get in the logs about where the warning came from.

Comment (chrisseaton) [reply src/main/java/org/truffleruby/RubyContext.java]: Well it means people don't know where to look to find the error. I won't object to a couple of cases, but we should not be in the habit of breaking this functionality of logging.

**Comment by nirvdrum** [comment]
There should be a finite list of things that need to be patched for pre-initialized contexts to work. I'm hopeful that if anything the list will shrink, not grow. I appreciate there's a risk it may get out of date (we have a similar problem with the docs on MRI parity, for instance), but I think it's still worth doing.

**Comment by eregon** [comment]
That list will keep changing and I have no interest to maintain the code + a copy in the documentation, it will inevitably get outdated. It's unlikely that the list will shrink, so far the more we do during pre-initialization, the more we need to delay/patch later as well.
What do you think of the general documentation about patching I added in the last commit?

**Comment by nirvdrum** [comment]
The docs look good. Thanks! I just want to avoid a situation where you're the only person that understands all the nuances of how pre-init contexts work.

**Pull request merged**